src/main/java/org/jenkinsci/plugins/ssegateway/EventHistoryStore.java

```diff
@@ -0,0 +1,297 @@
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.security.ACL;
+import org.apache.commons.io.FileUtils;
+import org.jenkins.pubsub.ChannelSubscriber;
+import org.jenkins.pubsub.Message;
+import org.jenkins.pubsub.PubsubBus;
+import org.jenkinsci.plugins.ssegateway.sse.EventDispatcher;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Channel event message history store.
+ * <p>
+ * Currently stores event history in files on disk, purging them
+ * as they go "stale" (after they expire).
+ * 
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+@Restricted(NoExternalUse.class)
+public final class EventHistoryStore {
+
+    private static final Logger LOGGER = Logger.getLogger(EventHistoryStore.class.getName());
+    
+    private static File historyRoot;
+    private static long expiresAfter = (1000 * 60); // default of 1 minutes
+    private static Map<String, File> channelDirs = new ConcurrentHashMap<>();
+    private static Timer autoExpireTimer;
+    
+    private static final Map<String, AtomicInteger> channelSubsCounters = new ConcurrentHashMap<>();
+    private static final Map<String, EventHistoryLogger> channelLoggers = new ConcurrentHashMap<>();
+
+    @SuppressFBWarnings(value = "LI_LAZY_INIT_STATIC", 
+                justification = "internal class (marked @Restricted NoExternalUse + package private methods) - need it this way for testing.")
+    static void setHistoryRoot(@Nonnull File historyRoot) throws IOException {
+        // In a non-test mode, we only allow setting of the historyRoot 
+        // once (during plugin init - see Endpoint class).
+        if (EventHistoryStore.historyRoot != null && !Util.isTestEnv()) {
+            LOGGER.log(Level.SEVERE, "Invalid attempt to change historyRoot after it has already been set. Ignoring.");
+            return;
+        }
+        
+        if (!historyRoot.exists()) {
+            if (!historyRoot.mkdirs()) {
+                throw new IOException(String.format("Unexpected error creating historyRoot dir %s. Check permissions etc.", historyRoot.getAbsolutePath()));
+            }
+        }
+        EventHistoryStore.historyRoot = historyRoot;
+    }
+
+    static void setExpiryMillis(long expiresAfterMillis) {
+        // In a non-test mode, we don't allow setting of the expiresAfter at all.
+        if (!Util.isTestEnv()) {
+            LOGGER.log(Level.SEVERE, "Invalid attempt to change expiresAfterMillis. Ignoring.");
+            return;
+        }
+        EventHistoryStore.expiresAfter = expiresAfterMillis;
+    }
+
+    public static void store(@Nonnull Message message) {
+        try {
+            String channelName = message.getChannelName();
+            String eventUUID = message.getEventUUID();
+            File channelDir = getChannelDir(channelName);
+            File eventFile = new File(channelDir, eventUUID + ".json");
```

Comment:
Is this stored specific to the session which has missed the messages somehow? Maybe I'm missing the bit that is somehow associating missed messages per *user* to the queue?